File: sections/76_suffix_arrays_lcp/problems/c_pattern_occurrence_queries/solution.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> suffix_array(const string &s) {
  int n = s.size();
  vector<int> sa(n), r(n), nr(n);
  iota(sa.begin(), sa.end(), 0);
  for (int i = 0; i < n; ++i)
    r[i] = (unsigned char)s[i];
  for (int k = 1;; k *= 2) {
    sort(sa.begin(), sa.end(), [&](int a, int b) {
      return pair{r[a], a + k < n ? r[a + k] : -1} <
             pair{r[b], b + k < n ? r[b + k] : -1};
    });
    nr[sa[0]] = 0;
    for (int i = 1; i < n; ++i) {
      int a = sa[i - 1], b = sa[i];
      nr[b] = nr[a] + (pair{r[a], a + k < n ? r[a + k] : -1} <
                       pair{r[b], b + k < n ? r[b + k] : -1});
    }
    r = nr;
    if (r[sa.back()] == n - 1)
      break;
  }
  return sa;
}
vector<int> lcp_array(const string &s, const vector<int> &sa) {
  int n = s.size(), h = 0;
  vector<int> rank(n), lcp(max(0, n - 1));
  for (int i = 0; i < n; ++i)
    rank[sa[i]] = i;
  for (int i = 0; i < n; ++i) {
    int q = rank[i];
    if (!q)
      continue;
    int j = sa[q - 1];
    while (i + h < n && j + h < n && s[i + h] == s[j + h])
      ++h;
    lcp[q - 1] = h;
    if (h)
      --h;
  }
  return lcp;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  string s;
  int q;
  cin >> s >> q;
  auto sa = suffix_array(s);
  while (q--) {
    string p;
    cin >> p;
    auto cmp = [&](int at) {
      int z = s.compare(at, p.size(), p);
      return z < 0 ? -1 : z > 0 ? 1 : 0;
    };
    int l = 0, r = sa.size();
    while (l < r) {
      int m = (l + r) / 2;
      if (cmp(sa[m]) < 0)
        l = m + 1;
      else
        r = m;
    }
    int first = l;
    l = 0;
    r = sa.size();
    while (l < r) {
      int m = (l + r) / 2;
      if (cmp(sa[m]) <= 0)
        l = m + 1;
      else
        r = m;
    }
    cout << l - first << '\n';
  }
}

File: sections/76_suffix_arrays_lcp/problems/c_pattern_occurrence_queries/solution.py
def suffix_array(s):
    n = len(s)
    sa = list(range(n))
    rank = list(map(ord, s))
    step = 1
    while True:
        sa.sort(
            key=lambda i: (rank[i], rank[i + step] if i + step < n else -1)
        )
        new_rank = [0] * n
        for j in range(1, n):
            a, b = sa[j - 1], sa[j]
            left = (rank[a], rank[a + step] if a + step < n else -1)
            right = (rank[b], rank[b + step] if b + step < n else -1)
            new_rank[b] = new_rank[a] + (left < right)
        rank = new_rank
        if rank[sa[-1]] == n - 1:
            return sa
        step *= 2


def lcp_array(s, sa):
    n = len(s)
    rank = [0] * n
    for i, start in enumerate(sa):
        rank[start] = i
    lcp = [0] * (n - 1)
    height = 0
    for i in range(n):
        place = rank[i]
        if place == 0:
            continue
        j = sa[place - 1]
        while (
            i + height < n
            and j + height < n
            and s[i + height] == s[j + height]
        ):
            height += 1
        lcp[place - 1] = height
        height = max(0, height - 1)
    return lcp


def main():
    import sys

    data = sys.stdin.buffer.read().split()
    s = data[0].decode()
    sa = suffix_array(s)
    output = []
    for raw in data[2:]:
        pattern = raw.decode()

        def compare(start):
            piece = s[start : start + len(pattern)]
            return (piece > pattern) - (piece < pattern)

        left, right = 0, len(sa)
        while left < right:
            middle = (left + right) // 2
            if compare(sa[middle]) < 0:
                left = middle + 1
            else:
                right = middle
        first = left
        left, right = 0, len(sa)
        while left < right:
            middle = (left + right) // 2
            if compare(sa[middle]) <= 0:
                left = middle + 1
            else:
                right = middle
        output.append(str(left - first))
    print("\n".join(output))


if __name__ == "__main__":
    main()
